### ISSUE TITLE
[prefer_const_constructors] does not work for all Flutter Icon's

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [dart-lang/linter#1592]
## Description

Coming from [StackOverflow](https://stackoverflow.com/q/56397809/6509751), it seems like [prefer_const_constructors](https://dart-lang.github.io/linter/lints/prefer_const_constructors.html) does not detect most instances of [`Icon`](https://github.com/flutter/flutter/blob/bfa1d25bf9f51f19dec62e1f8e9c25e6ec6cf894/packages/flutter/lib/src/widgets/icon.dart#L51).

## Reproduction

For me personally, it does seem to work for **some** instances. I successfully got hints for `Icon(Icons.delete_sweep)`, `Icon(Icons.cancel)` and `Icon(Icons.send)`, however, there is no apparent reason for this behavior since [all constants in `Icons` are assigned the same way](https://github.com/flutter/flutter/blob/bfa1d25bf9f51f19dec62e1f8e9c25e6ec6cf894/packages/flutter/lib/src/material/icons.dart#L37).  
For the OP on StackOverflow, there are no warning at all for instances of `Icon`.

```dart
main() {
  runApp(WidgetsApp(home: Icon(Icons.add), color: Color(0xffffffff)));
}
```
This warns me for the instance of `Color` but not for the instance of `Icon`, even though it should warn me for both.

No warnings like this:

```dart
runApp(WidgetsApp(home: Icon(Icons.add), color: const Color(0xffffffff)));
```

**Comment by sgon00** [comment]
Adding Flutter and Dart versions:

```
[✓] Flutter (Channel master, v1.6.1-pre.47, on Mac OS X 10.13.6 17G4015, locale en-CN)
    • Flutter version 1.6.1-pre.47 at /Users/user/opt/flutter
    • Framework revision 40300fb8a9 (11 days ago), 2019-05-20 12:04:21 -0400
    • Engine revision 3a509192aa
    • Dart version 2.3.1 (build 2.3.1-dev.0.0 ed169c4bb0)
```

**Comment by pq** [comment]
Thanks for the report!  Could you add a snippet that reproduces the issue?  Happy to take a look. 👍

**Comment by jamesderlin** [comment]
The sample app generated by `flutter create` demonstrates this.

**Comment by creativecreatorormaybenot** [comment]
@pq 

```dart
main() {
  runApp(WidgetsApp(home: Icon(Icons.add), color: Color(0xffffffff)));
}
```
This warns me for the instance of `Color` but not for the instance of `Icon`, even though it should warn me for both.

**Comment by sgon00** [comment]
Hi, I found a new case that `prefer_const_constructors` doesn't work. I am not sure if I should open a new bug for this or not. 

The complete runnable code:

```dart
import 'package:flutter/material.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'test',
      theme: ThemeData(
        primarySwatch: Colors.blue
      ),
      home: MyWidget()
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: MyConstWidget(color: Colors.black),
    );
  }
}

class MyConstWidget extends StatelessWidget {

  const MyConstWidget({
    Key key,
    @required this.color,
  }) : super(key: key);

  final Color color;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Center(child: Text('Hello World', style: TextStyle(color: color)))
    );
  }

}
```

`const MyConstWidget(color: Colors.black),` works too, but there is no warning for `MyConstWidget(color: Colors.black),`.

Thanks a lot.

**Comment by a14n** [comment]
This issue seems related to analysis server. Here's a simple repro:

```dart
// -----------   a.dart
import 'b.dart';

void main() {
  A(B.k);
}

class A {
  const A(p);
}


// -----------   b.dart
class B {
  static const k = '';
}
```

`flutter analyze` (that uses analysis server) does not trigger `prefer_const_constructor` but with linter directly there's correctly a lint:

```
$ dart $LINTER_HOME/bin/linter.dart --packages .packages --rules prefer_const_constructors lib/a.dart 
[...]/test/lib/a.dart 4:3 [lint] Prefer const with constant constructors.
  A(B.k);
  ^^^^^^

1 file analyzed, 1 issue found, in 1363 ms.
```

/cc @scheglov

**Comment by creativecreatorormaybenot** [comment]
@pq How can I make sure that my version of linter has this pull request?

I am asking because I still see the same issue, i.e. it is not fixed for me. I am on the latest Flutter master channel version.

**Comment by pq** [comment]
Yeah.  This is more difficult to figure out than it should be.  😬 

That fix made it into linter `0.1.91` which got into Dart as of `2.3.3-dev.0.0`.  You should be able to see what version of Dart you're on by running `flutter doctor -v`...

**Comment by creativecreatorormaybenot** [comment]
@pq Well, I am on Dart version `2.4.0`.

**Comment by srawlins** [comment]
Oof, sorry @creativecreatorormaybenot you've hit sort of a seam in Dart SDK versions. 2.4.0-dev.0.0 has very few changes past 2.3.3-dev.0.0 ([git comparison](https://github.com/dart-lang/sdk/compare/2.3.3-dev.0.0...2.4.0-dev.0.0)), so you should have linter 0.1.91.
